stan/math/opencl/kernel_generator/reduction_2d.hpp

```diff
@@ -0,0 +1,342 @@
+   */
+  inline kernel_parts generate(const std::string& row_index_name,
+                               const std::string& col_index_name,
+                               const bool view_handled,
+                               const std::string& var_name_arg) const {
+    kernel_parts res;
+    res.declarations = "__local " + type_str<Scalar>() + " " + var_name_
+                       + "_local[LOCAL_SIZE_];\n" + type_str<Scalar>() + " "
+                       + var_name_ + " = " + init_ + ";\n";
+    res.body = var_name_ + " = " + Operation::generate(var_name_, var_name_arg)
+               + ";\n";
+    res.reduction_2d =
+          var_name_ + "_local[lid_i] = " + var_name_ + ";\n"
+          "barrier(CLK_LOCAL_MEM_FENCE);\n"
+          "for (int step = lsize_i / REDUCTION_STEP_SIZE; "
+                "step > 0; step /= REDUCTION_STEP_SIZE) {\n"
+          "  if (lid_i < step) {\n"
+          "    for (int i = 1; i < REDUCTION_STEP_SIZE; i++) {\n"
+          "      " + var_name_ + "_local[lid_i] = " +
+        Operation::generate(var_name_ + "_local[lid_i]",
+                            var_name_ + "_local[lid_i + step * i]") + ";\n"
+          "    }\n"
+          "  }\n"
+          "  barrier(CLK_LOCAL_MEM_FENCE);\n"
+          "}\n";
+    return res;
+  }
+
+  /**
+   * Number of rows of a matrix that would be the result of evaluating this
+   * expression.
+   * @return number of rows
+   */
+  inline int rows() const {
+    int arg_rows = this->template get_arg<0>().rows();
+    int arg_cols = this->template get_arg<0>().cols();
+    if (arg_cols == 0) {
+      return 1;
+    }
+    if (arg_cols == -1) {
+      return -1;
+    }
+    return internal::colwise_reduction_wgs_rows(arg_rows, arg_cols);
+  }
+
+  /**
+   * Number of columns of a matrix that would be the result of evaluating this
+   * expression.
+   * @return number of columns
+   */
+  inline int cols() const {
+    int arg_rows = this->template get_arg<0>().rows();
+    int arg_cols = this->template get_arg<0>().cols();
+    if (arg_cols == 0) {
+      return 0;
+    }
+    if (arg_cols == -1) {
+      return -1;
+    }
+    int wgs_rows = internal::colwise_reduction_wgs_rows(arg_rows, arg_cols);
+    if (wgs_rows == 0) {
+      return 0;
+    }
+    return (arg_cols + wgs_rows - 1) / wgs_rows;
+  }
+
+  /**
+   * Number of rows threads need to be launched for.
+   * @return number of rows
+   */
+  inline int thread_rows() const { return this->template get_arg<0>().rows(); }
+
+  /**
+   * Number of rows threads need to be launched for.
+   * @return number of rows
+   */
+  inline int thread_cols() const { return this->template get_arg<0>().cols(); }
+
+  /**
+   * Determine indices of extreme sub- and superdiagonals written.
+   * @return pair of indices - bottom and top diagonal
+   */
+  inline std::pair<int, int> extreme_diagonals() const {
+    return {-rows() + 1, cols() - 1};
+  }
+};  // namespace math
```

Comment:
[optional] you could tag a bunch of these as noexcept as well but no biggie on that